eth/backend.go

```diff
@@ -0,0 +1,171 @@
+package eth
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethabi "github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+)
+
+var abis = []string{
+	contracts.BondingManagerABI,
+	contracts.ControllerABI,
+	contracts.LivepeerTokenABI,
+	contracts.LivepeerTokenFaucetABI,
+	contracts.MinterABI,
+	contracts.RoundsManagerABI,
+	contracts.ServiceRegistryABI,
+	contracts.TicketBrokerABI,
+}
+
+type Backend interface {
+	ethereum.ChainReader
+	ethereum.ChainStateReader
+	ethereum.ChainSyncReader
+	ethereum.TransactionReader
+	ethereum.TransactionSender
+	ethereum.ContractCaller
+	ethereum.PendingContractCaller
+	ethereum.PendingStateReader
+	// ethereum.PendingStateEventer
+	ethereum.GasEstimator
+	ethereum.GasPricer
+	ethereum.LogFilterer
+
+	NetworkID(ctx context.Context) (*big.Int, error)
+}
+
+type backend struct {
+	*ethclient.Client
+	methods      map[string]*method
+	nonceManager *NonceManager
+}
+
+type method struct {
+	name string
+	abi  *abi.ABI
+}
+
+func NewBackend(client *ethclient.Client) (Backend, error) {
+	methods, err := makeMethodsMap()
+	if err != nil {
+		return nil, err
+	}
+
+	return &backend{
+		client,
+		methods,
+		NewNonceManager(client),
+	}, nil
+}
+
+func (b *backend) PendingNonceAt(ctx context.Context, account common.Address) (uint64, error) {
+	var nonce uint64
+	var err error
+	if b.nonceManager != nil {
+		b.nonceManager.Lock(account)
+		defer func() {
+			if err == nil {
+				b.nonceManager.Update(account, nonce)
```

Comment:
Actually we are doing it already but it might not be very clear.

`BoundContract.Transact` will call `Backend.PendingNonceAt()` before calling `Backend.SendTransaction()`.

Reply:
Hm so the difference now is that the `defer` statement is going to execute when `Backend.PendingNonceAt()` returns. Previously, the `defer` statement executed when `BoundContract.Transact()` returned and would check the last tracked error in the method which happened to be the one returned by `SendTransaction()`. So, the go-ethereum fork *was* checking if `SendTransaction()` was successful although the approach used was definitely confusing.

Reply:
Ah yes I see, the order of declaration there is a little confusing.

Reply:
The issue is that `SendTransaction` doesn't really have access to the sender address to call `Update` on `NonceManager`

Reply:
A workaround is to recover the sender address from the signed transaction that is passed into `SendTransaction()`.

We can use the [tx.AsMessage()](https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go#L220) method to return a [Message](https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go#L412) type and then we can call `.From()` to return the account address that signed the tx. See this [SO post](https://ethereum.stackexchange.com/questions/67449/given-a-raw-transaction-how-to-get-senders-address) for example usage.

For now we'll want to use `types.HomesteadSigner{}` as the provided signer since this is the hardcoded signer used by [BoundContract](https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/bind/base.go#L237). Side note: `BoundContract` should be using the EIP155Signer for signing transactions...there is an [open go-ethereum PR](https://github.com/ethereum/go-ethereum/pull/17046) with this change.